### PR TITLE
test: cover per-EIP config gating of frame transactions

### PR DESCRIPTION
**Pull request by AnkushinDaniil**
Adds regression coverage that each frame-family EIP (8141, 8250, 8272, 7906, 7805) is independently gated by its chain-config transition timestamp, so a single build can join a devnet running any combination.

## Changes
- **TxValidator**: reject a type `0x06` transaction when EIP-8141 is disabled.
- **FrameTxValidation**: reject frame mode 5 while its owning EIP is inactive.
- **FrameTxProcessor**: opcode bytes match the family registry; an unallocated `0xB0-0xBF` byte exceptionally halts; single-config orthogonality checks that the keyed-nonce, recent-root and post-tx envelopes each run only under their own fork.
- **FrameTxBlockGas**: a frame transaction starts owing state gas only after activation.
- **ChainSpecBasedSpecProvider**: each frame-family EIP flips on exactly at its own transition timestamp.
- **Engine capabilities**: the FOCIL engine methods advertise only when EIP-7805 is enabled.
- **Optimism**: guard a nullable `Logs` dereference in the deposit builder so `Nethermind.Merge.Plugin.Test` builds.

Each control was verified by temporarily reverting its guard and confirming the test fails.

## Types of changes
- [x] Test
- [x] Bugfix (Optimism build)

## Testing
Affected suites pass: Evm.Test (16), Blockchain.Test, Core.Test, Specs.Test, Merge.Plugin.Test (12).

### CODE REVIEW

**Review by wurdum — APPROVED** [review]
All issues I've found are already reported here 😄

File: src/Nethermind/Nethermind.Optimism/CL/Derivation/DepositTransactionBuilder.cs [hunk 1]
```diff
@@ -78,7 +78,7 @@ public IEnumerable<Transaction> BuildUserDepositTransactions(ReceiptForRpc[] rec
         foreach (ReceiptForRpc receipt in receipts)
         {
             if (receipt.Status != StatusCode.Success) continue;
-            foreach (LogEntryForRpc log in receipt.Logs)
+            foreach (LogEntryForRpc log in receipt.Logs ?? [])
```
Comment (flcl42) [inline src/Nethermind/Nethermind.Optimism/CL/Derivation/DepositTransactionBuilder.cs:82]: **[MEDIUM] Null Logs from the L1 receipts RPC are now silently treated as no deposits**

BuildUserDepositTransactions consumes ReceiptForRpc[] fetched from an external L1 node via eth_getBlockReceipts (EthereumL1Bridge.GetReceiptsByBlockHash -> EthereumEthApi.GetReceiptsByHash), so the receipts are untrusted JSON-RPC input and ReceiptForRpc.Logs is nullable (LogEntryForRpc[]?). Before this change, a receipt whose JSON omitted or nulled the logs field caused a NullReferenceException at this line — a loud failure during derivation. With `receipt.Logs ?? []` the same malformed receipt is now silently treated as containing no logs, so any TransactionDeposited events in it are dropped and the derived OptimismPayloadAttributes omit deposit transactions that every correctly deriving node includes, producing a divergent L2 payload instead of a visible error. The change is also unrelated to the PR's stated test-only scope and carries no test covering the null-Logs path. Retaining a loud rejection (or an explicit validation error) for a null Logs field would preserve the fail-fast behavior on malformed L1 responses.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxBlockGasTests.cs [hunk 2]
```diff
@@ -295,6 +297,57 @@ public void Execute_PostTxOutcome_DecidesWhetherTheBodyOwesStateGas(bool asserti
...
+        Deploy(Sender, ApproveCode(TxFrame.ApproveExecutionAndPayment), UInt256.Parse("100000000000000000000"));
+        Deploy(Writer, Prepare.EvmCode.PushData(1).PushData(0).Op(Instruction.SSTORE).Op(Instruction.STOP).Done);
+        Deploy(laterWriter, Prepare.EvmCode.PushData(1).PushData(0).Op(Instruction.SSTORE).Op(Instruction.STOP).Done);
+
+        TestSpecProvider provider = (TestSpecProvider)_specProvider;
+        OverridableReleaseSpec specAfterFork = (OverridableReleaseSpec)provider.GenesisSpec;
+        OverridableReleaseSpec specBeforeFork = new(Eip8141Prototype.Instance) { IsEip7906Enabled = true, IsEip8037Enabled = false };
+
+        Transaction sample = FrameTx(nonce: 7, laterWriter);
+        Hash256 hashAfterFork = ConsensusHash(sample);
```
Comment (flcl42) [inline src/Nethermind/Nethermind.Evm.Test/FrameTxBlockGasTests.cs]: **[LOW] Fork-boundary consensus-hash assertion cannot fail because ConsensusHash ignores the spec**

In Execute_WritingFrameTxAcrossTheFrameLimitsBoundary_KeepsTheConsensusHashAndMetersStateOnlyAfterActivation, the test swaps provider.GenesisSpec between specAfterFork and specBeforeFork around the two ConsensusHash(sample) calls and asserts hashAfterFork == hashBeforeFork with the message that re-encoding keeps the consensus hash in both eras. ConsensusHash, however, only RLP-encodes the transaction via TxDecoder.Instance with RlpBehaviors.SkipTypedWrapping and never reads the spec provider, so both hashes are computed from the identical sample object under identical encoding behavior; the assertion is trivially true and the two GenesisSpec assignments around it are inert. If a future change made the frame envelope encoding fork-dependent (for example dropping the state dimension before frameLimitsTime), this test would still pass while claiming to pin the cross-fork hash stability. Anchoring the hash to the fork era the test actually executes under, or asserting against a literal expected hash, would give the check discriminating power.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs [hunk 3]
```diff
@@ -2416,6 +2416,28 @@ public void Execute_RecentRootReferencesBeforeTheReferenceFork_AreRejected()
...
+        DeploySmartSender(ApproveCode(TxFrame.ApproveExecutionAndPayment));
+        DeployContract(Recipient, Prepare.EvmCode.Op(Instruction.STOP).Done);
+
+        Transaction postTx = FrameTx(nonce: 0, SelfVerifyFrame(), Frame(TxFrame.ModePostTx, target: Recipient));
+        Transaction keyed = FrameTx(nonce: 1, SelfVerifyFrame());
+        keyed.NonceKeys = [(UInt256)7];
+        Transaction rooted = FrameTx(nonce: 1, SelfVerifyFrame());
+        rooted.RecentRootReferences = [];
+
+        using (Assert.EnterMultipleScope())
```
Comment (flcl42) [inline src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:2615]: **[LOW] Assert.EnterMultipleScope wraps three state-mutating transaction executions**

In Execute_AssertionForkWithoutKeyedNoncesOrRecentRoots_RunsPostTxButRefusesTheOtherEnvelopes, the three Process calls each execute a transaction against the shared _stateProvider and are placed inside a single Assert.EnterMultipleScope. NUnit evaluates every assertion in the scope even after a failure, so if the postTx execution assertion fails, the keyed and rooted transactions still run against a state where the sender nonce was never advanced by postTx; their nonce-1 envelopes then get rejected for the wrong reason while the Is.False assertions still pass, masking the real regression. Keeping the executions outside the multiple scope, or asserting on pre-collected results, would preserve the stop-on-first-failure behavior the project test guidelines describe for state-mutating calls.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs [hunk 4]
```diff
@@ -3194,4 +3216,72 @@ public ulong Execute_ReferenceCountTxParam_IsGatedOnTheFork(bool beforeTheFork)
...
+    [TestCase(Instruction.FRAMEDATALOAD, (byte)0xB1, TestName = "RegistryByte_FRAMEDATALOAD_0xB1")]
+    [TestCase(Instruction.FRAMEDATACOPY, (byte)0xB2, TestName = "RegistryByte_FRAMEDATACOPY_0xB2")]
+    [TestCase(Instruction.FRAMEPARAM, (byte)0xB3, TestName = "RegistryByte_FRAMEPARAM_0xB3")]
+    [TestCase(Instruction.SIGPARAM, (byte)0xB4, TestName = "RegistryByte_SIGPARAM_0xB4")]
+    [TestCase(Instruction.SIGDATACOPY, (byte)0xB5, TestName = "RegistryByte_SIGDATACOPY_0xB5")]
+    [TestCase(Instruction.RECENTROOTREFLOAD, (byte)0xB6, TestName = "RegistryByte_RECENTROOTREFLOAD_0xB6")]
+    public void FrameOpcodeByte_MatchesTheSpecRegistry(Instruction opcode, byte registryByte)
+        => Assert.That((byte)opcode, Is.EqualTo(registryByte));
+
+    [TestCase((byte)0xBA, TestName = "UnallocatedFrameOpcode_0xBA_Halts")]
```
Comment (flcl42) [inline src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:3548]: **[LOW] Unallocated frame-opcode test covers only four of six gap bytes**

The gap between EIP-8141/EIP-8272 and EIP-7906 opcodes is 0xBA-0xBF (six values). The new parameterized test at src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:3231 enumerates 0xBA, 0xBB, 0xBD, 0xBF and omits 0xBC and 0xBE. The underlying dispatch table in EvmInstructions.cs leaves all six entries null, but a regression that accidentally mapped 0xBC or 0xBE to a valid handler would not be caught. The failure path is a future opcode-table change that enables one of the missing bytes; the test would still pass while the VM no longer exceptionally halts for that byte. Coverage of the full 0xBA-0xBF range would give the test discriminating power.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs [hunk 5]
```diff
@@ -3194,4 +3216,72 @@ public ulong Execute_ReferenceCountTxParam_IsGatedOnTheFork(bool beforeTheFork)
...
+        DeploySmartSender(ApproveCode(TxFrame.ApproveExecutionAndPayment));
+        DeployContract(Recipient, Prepare.EvmCode.Op(Instruction.STOP).Done);
+
+        Transaction keyed = FrameTx(nonce: 0, SelfVerifyFrame());
+        keyed.NonceKeys = [1, 7];
+        Assert.That(Process(keyed).TransactionExecuted, Is.True);
+
+        Transaction referencing = FrameTx(nonce: 0, SelfVerifyFrame());
+        referencing.RecentRootReferences = [];
+        Assert.That(Process(referencing, slotNumber: HeadSlot).TransactionExecuted, Is.False);
```
Comment (flcl42) [inline src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs]: **[LOW] Fork-isolation assertions verify only execution result, not rejection cause**

Execute_KeyedNonceForkWithoutReferencesOrAssertions_RunsKeyedAndRefusesTheOtherEnvelopes and the symmetric reference test at src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:3262 assert only TransactionExecuted Is.False for the two shapes that must be refused before their fork (RecentRootReferences when IsEip8272Enabled=false and POST_TX when IsEip7906Enabled=false). No Error, ErrorType, or ErrorDescription is checked. On the current code the processor does return MalformedTransaction with "not enabled", but a different failure (nonce mismatch, malformed envelope, gas) would also satisfy Is.False, so the test would still pass while the fork gate is not the reason for refusal. Assertions that also verify MalformedTransaction and the "not enabled" detail would distinguish fork gating from other validation paths.

**Comment by AnkushinDaniil** [comment]
Rebased onto the current base; the diff is now test-only. The `DepositTransactionBuilder` change is dropped: #13021 landed the equivalent `receipt.Logs ?? []` guard on the base branch, so the null-logs handling the MEDIUM flagged now lives there rather than here. That also settles the concern that this PR silently coerced null logs, since the coercion is no longer part of it.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxBlockGasTests.cs [hunk 6]
```diff
@@ -295,6 +297,57 @@ public void Execute_PostTxOutcome_DecidesWhetherTheBodyOwesStateGas(bool asserti
...
+        provider.GenesisSpec = specBeforeFork;
+        Hash256 hashBeforeFork = ConsensusHash(sample);
+        provider.GenesisSpec = specAfterFork;
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxBlockGasTests.cs]: Fixed in 50de456. Dropped the vacuous consensus-hash assertion and its dead helper; the test is renamed to only claim the state metering it actually verifies.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxBlockGasTests.cs (hunk 2 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxBlockGasTests.cs]: Fixed in 50de456. Removed the inert spec swaps and the tautological equality; the test now asserts only the state metering, which is the real coverage.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs [hunk 7]
```diff
@@ -2416,6 +2416,28 @@ public void Execute_RecentRootReferencesBeforeTheReferenceFork_AreRejected()
...
+        Transaction keyed = FrameTx(nonce: 1, SelfVerifyFrame());
+        keyed.NonceKeys = [(UInt256)7];
+        Transaction rooted = FrameTx(nonce: 1, SelfVerifyFrame());
+        rooted.RecentRootReferences = [];
+
+        using (Assert.EnterMultipleScope())
+        {
+            Assert.That(Process(postTx).TransactionExecuted, Is.True, "the assertion fork must run a POST_TX frame");
+            Assert.That(Process(keyed).TransactionExecuted, Is.False, "keyed nonces must stay refused before their fork");
+            Assert.That(Process(rooted, slotNumber: 1_001).TransactionExecuted, Is.False, "recent-root references must stay refused before their fork");
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs]: Fixed in 50de456. The fork-gated negative legs now assert Error is MalformedTransaction and ErrorDescription carries the not-enabled detail, so a nonce or gas refusal no longer passes for the gate.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs (hunk 5 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs]: Fixed in 50de456. Both fork-isolation tests now assert the not-enabled ErrorDescription (KeyedNoncesNotEnabled / RecentRootReferencesNotEnabled / PostTxNotEnabled) on each refused leg.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs (hunk 3 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:2615]: Fixed in 50de456. The three executions now run before the scope and only the collected results are asserted inside it, restoring stop-on-first-failure.

File: src/Nethermind/Nethermind.Core.Test/FrameTxValidationTests.cs [hunk 8]
```diff
@@ -51,6 +51,8 @@ private static IEnumerable<TestCaseData> ConstraintCases()
         // assert frame.mode < 4 (EIP-7906 widened it from 3)
         yield return Case("FrameModeFour_InvalidMode",
             static tx => tx.Frames = [Frame(mode: 4)], FrameTxValidation.InvalidMode);
+        yield return Case("FrameModeFive_InvalidMode",
+            static tx => tx.Frames = [Frame(mode: 5)], FrameTxValidation.InvalidMode);
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Core.Test/FrameTxValidationTests.cs]: Fixed in 50de456. Dropped the redundant mode-5 case and added a parameterized case that pins the postTxEnabled gate to PostTxNotEnabled.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs [hunk 9]
```diff
@@ -3194,4 +3216,72 @@ public ulong Execute_ReferenceCountTxParam_IsGatedOnTheFork(bool beforeTheFork)
...
+
+    [TestCase(Instruction.APPROVE, (byte)0xAA, TestName = "RegistryByte_APPROVE_0xAA")]
+    [TestCase(Instruction.TXPARAM, (byte)0xB0, TestName = "RegistryByte_TXPARAM_0xB0")]
+    [TestCase(Instruction.FRAMEDATALOAD, (byte)0xB1, TestName = "RegistryByte_FRAMEDATALOAD_0xB1")]
+    [TestCase(Instruction.FRAMEDATACOPY, (byte)0xB2, TestName = "RegistryByte_FRAMEDATACOPY_0xB2")]
+    [TestCase(Instruction.FRAMEPARAM, (byte)0xB3, TestName = "RegistryByte_FRAMEPARAM_0xB3")]
+    [TestCase(Instruction.SIGPARAM, (byte)0xB4, TestName = "RegistryByte_SIGPARAM_0xB4")]
+    [TestCase(Instruction.SIGDATACOPY, (byte)0xB5, TestName = "RegistryByte_SIGDATACOPY_0xB5")]
+    [TestCase(Instruction.RECENTROOTREFLOAD, (byte)0xB6, TestName = "RegistryByte_RECENTROOTREFLOAD_0xB6")]
+    public void FrameOpcodeByte_MatchesTheSpecRegistry(Instruction opcode, byte registryByte)
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:3545]: Fixed in 50de456. Added TXTRACE 0xB7, TXDIFF 0xB8, EVENTDATACOPY 0xB9 to the registry test \u2014 the three that deviate from EIP-7906's draft bytes.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs [hunk 10]
```diff
@@ -3194,4 +3216,72 @@ public ulong Execute_ReferenceCountTxParam_IsGatedOnTheFork(bool beforeTheFork)
...
+    {
+        DeploySmartSender(ApproveCode(TxFrame.ApproveExecutionAndPayment));
+        DeployContract(Observer, Prepare.EvmCode
+            .Op(opcode)
+            .PushData(1).PushData(0).Op(Instruction.SSTORE).Op(Instruction.STOP).Done);
+
+        Transaction tx = FrameTx(nonce: 0, SelfVerifyFrame(), Frame(TxFrame.ModeDefault, target: Observer));
+
+        Assert.That(Process(tx, slotNumber: HeadSlot).TransactionExecuted, Is.True);
+        AssertStorage(Observer, 0, UInt256.Zero);
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:3564]: Fixed in 50de456. Added 0xBC and 0xBE, so the unallocated test now covers all six of 0xBA-0xBF. On the halt-vs-revert point: the storage-not-committed assert stays; the frame's gas is metered by the outer loop, so a dedicated gas-consumed assert here would need the inner tracer \u2014 noting it rather than expanding scope.

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs (hunk 4 above)
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs:3548]: Fixed in 50de456. The unallocated test now enumerates the full 0xBA-0xBF range (added 0xBC and 0xBE).

File: src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs [hunk 11]
```diff
@@ -2416,6 +2416,28 @@ public void Execute_RecentRootReferencesBeforeTheReferenceFork_AreRejected()
         Assert.That(result.ErrorDescription, Does.Contain("not enabled"));
     }
 
+    [Test]
+    public void Execute_AssertionForkWithoutKeyedNoncesOrRecentRoots_RunsPostTxButRefusesTheOtherEnvelopes()
+    {
+        ((TestSpecProvider)_specProvider).GenesisSpec =
+            new OverridableReleaseSpec(Eip8141Prototype.Instance) { IsEip8250Enabled = false, IsEip8272Enabled = false, IsEip7906Enabled = true };
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Evm.Test/FrameTxProcessorTests.cs]: Fixed in 50de456. Now mutates the fixture _spec (IsEip8250Enabled/IsEip8272Enabled = false) instead of rebuilding a replacement, matching the sibling tests and leaving _spec authoritative.

File: src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.Capabilities.cs [hunk 12]
```diff
@@ -25,4 +25,19 @@ public void Engine_getBlobsV4_capability_follows_eip7843(bool eip7843Enabled, bo
...
+    [TestCase(nameof(IEngineRpcModule.engine_getInclusionListV1))]
+    [TestCase(nameof(IEngineRpcModule.engine_forkchoiceUpdatedV5))]
+    [TestCase(nameof(IEngineRpcModule.engine_newPayloadWithWitnessV6))]
+    public void Engine_inclusionList_capabilities_follow_eip7805(string method)
+    {
+        IReleaseSpec enabledSpec = new ReleaseSpec { IsEip7805Enabled = true };
+        IReleaseSpec disabledSpec = new ReleaseSpec { IsEip7805Enabled = false };
+        EngineRpcCapabilitiesProvider enabledProvider = new(new TestSingleReleaseSpecProvider(enabledSpec));
+        EngineRpcCapabilitiesProvider disabledProvider = new(new TestSingleReleaseSpecProvider(disabledSpec));
+
```
[the remark this answers is not in the document]
Comment (AnkushinDaniil) [reply src/Nethermind/Nethermind.Merge.Plugin.Test/EngineModuleTests.Capabilities.cs:39]: Fixed in 50de456. Wrapped both capability asserts in a multiple-scope so the disabled-leg failure is reported too.
